calyx/src/passes/sharing_components.rs

```diff
@@ -109,52 +103,119 @@ impl<T: ShareComponents> Visitor for T {
+                for conflict in node_conflicts {
+                    live_at_node
+                        .entry(&id_to_type[conflict])
+                        .or_default()
+                        .insert(conflict);
+                }
+            }
+        }
+
+        // Closure so that we can take a group/invoke, and get all of the cells live
+        // at that group/invoke, *organized by type* in the form of a HashMap.
+        let lookup_conflicts_by_type =
+            |node: &ir::Id| -> &HashMap<&ir::CellType, HashSet<&ir::Id>> {
+                node_by_type_map.get(node).unwrap_or_else(|| {
+                    unreachable!("no node conflict map for {}", node)
+                })
+            };
+
+        // conflict (a,b) is in par_conflicts if a and b run in parallel w/ each other
         let par_conflicts = ScheduleConflicts::from(&*comp.control.borrow());
-        let node_conflicts = par_conflicts
+
+        // Building node_conflicts,which is a map from nodes to another map.
+        // nodes are inovkes/enables. maps are the cells live at nodes that may be run in
+        // parrallel with node, and these cells are again organized by cell type.
+        let mut node_conflicts = par_conflicts
             .all_conflicts()
             .into_grouping_map_by(|(g1, _)| g1.clone())
             .fold(
-                HashMap::new(),
-                |mut acc: HashMap<ir::CellType, Vec<ir::Id>>,
-                 _,
-                 (_, conflicted_group)| {
-                    for conflict in
-                        self.lookup_node_conflicts(&conflicted_group)
-                    {
-                        acc.entry(id_to_type[conflict].clone())
-                            .or_default()
-                            .push(conflict.clone())
+                HashMap::<&ir::CellType, HashSet<&ir::Id>>::new(),
+                |mut acc, _, (_, conflicted_group)| {
+                    let new_conflicts =
+                        lookup_conflicts_by_type(&conflicted_group);
+                    for (cell_type, nodes) in new_conflicts {
+                        acc.entry(cell_type).or_default().extend(nodes);
                     }
                     acc
                 },
             );
 
-        node_conflicts
-            .into_iter()
-            .for_each(|(node, conflict_group_b)| {
-                for a in self.lookup_node_conflicts(&node) {
-                    let g = graphs_by_type.get_mut(&id_to_type[a]).unwrap();
-                    if let Some(confs) = conflict_group_b.get(&id_to_type[a]) {
-                        for b in confs {
-                            if a != b {
-                                g.insert_conflict(a, b);
-                            }
+        // add conflicts
+        for node_name in &invokes_enables {
+            let node_confs_by_type = lookup_conflicts_by_type(node_name);
+            match node_conflicts.get_mut(node_name) {
```

Comment:
Should this be an `if let`?

Reply:
I am looking at the Some(_) case as well. The way git shows the changes are weird though. But the `Some(conflict_map) ` is the `Some` pattern.